### PR TITLE
Initial porting of SignalR SqlServer

### PR DESCRIPTION
**Pull request by Xiaohongt**
Note:
1. In full clr, we  still use SqlDependency.
2. In core clr, we only use pulling to detect new data in DB now, wait for the final decision if core clr will support low level SqlNotificationRequest or not.
3. In core clr, some interface types for Sql Client also are  removed, and changed to use some class types instead.

### CODE REVIEW

File: src/Microsoft.AspNet.SignalR.SqlServer/SqlServerServiceCollectionExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+using Microsoft.AspNet.SignalR;
+using Microsoft.AspNet.SignalR.Messaging;
+using Microsoft.AspNet.SignalR.SqlServer;
+using Microsoft.Framework.ConfigurationModel;
+
+namespace Microsoft.Framework.DependencyInjection
+{
+    public static class SqlServerServiceCollectionExtensions
+    {
+        public static IServiceCollection AddSignalRSqlServer(this IServiceCollection services, Action<SqlScaleoutConfiguration> configureOptions = null)
```
Comment (BrennanConroy) [inline src/Microsoft.AspNet.SignalR.SqlServer/SqlServerServiceCollectionExtensions.cs]: I think we decided this would be AddSqlServer?

File: src/Microsoft.AspNet.SignalR.SqlServer/TaskAsyncHelper.cs [hunk 2]
```diff
@@ -0,0 +1,1194 @@
+// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
```
Comment (BrennanConroy) [inline src/Microsoft.AspNet.SignalR.SqlServer/TaskAsyncHelper.cs]: Is this file actually used? Last I checked there were only a couple simple uses of the TaskAsyncHelper that should be easy to replace.

Comment (halter73) [inline src/Microsoft.AspNet.SignalR.SqlServer/TaskAsyncHelper.cs]: :+1: for removing `TaskAsyncHelper` from this project.

File: src/Microsoft.AspNet.SignalR.SqlServer/SqlScaleoutConfiguration.cs [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+using System;
+using Microsoft.AspNet.SignalR.Messaging;
+
+namespace Microsoft.AspNet.SignalR
+{
+    /// <summary>
+    /// Settings for the SQL Server scale-out message bus implementation.
+    /// </summary>
+    public class SqlScaleoutConfiguration : ScaleoutConfiguration
```
Comment (halter73) [inline src/Microsoft.AspNet.SignalR.SqlServer/SqlScaleoutConfiguration.cs:12]: We should Rename `SqlScaleoutConfiguration` and `ScaloutConfiguration` to `SqlScaleoutOptions` and `ScaloutOptions` respectively.

File: test/Microsoft.AspNet.SignalR.SqlServer.Tests/Properties/AssemblyInfo.cs [hunk 4]
```diff
@@ -0,0 +1,36 @@
+﻿using System.Reflection;
```
Comment (halter73) [inline test/Microsoft.AspNet.SignalR.SqlServer.Tests/Properties/AssemblyInfo.cs]: We don't need this file anymore.

File: src/Microsoft.AspNet.SignalR.SqlServer/Properties/AssemblyInfo.cs [hunk 5]
```diff
@@ -0,0 +1,13 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Reflection;
+using System.Runtime.CompilerServices;
+
+[assembly: AssemblyTitle("Microsoft.AspNet.SignalR.SqlServer")]
+[assembly: AssemblyDescription("Assembly containing components for scaling-out SignalR applications using SQL Server.")]
+#if SIGNED
+[assembly: InternalsVisibleTo("Microsoft.AspNet.SignalR.SqlServer.Tests, PublicKey=0024000004800000940000000602000000240000525341310004000001000100b5fc90e7027f67871e773a8fde8938c81dd402ba65b9201d60593e96c492651e889cc13f1415ebb53fac1131ae0bd333c5ee6021672d9718ea31a8aebd0da0072f25d87dba6fc90ffd598ed4da35e44c398c454307e8e33b8426143daec9f596836f97c8f74750e5975c64e2189f45def46b2a2b1247adc3652bf5c308055da9")]
```
Comment (halter73) [inline src/Microsoft.AspNet.SignalR.SqlServer/Properties/AssemblyInfo.cs]: We aren't strong naming. No need for the PublicKey or this SIGNED section.

File: src/Microsoft.AspNet.SignalR.SqlServer/Properties/AssemblyInfo.cs [hunk 6]
```diff
@@ -0,0 +1,13 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Reflection;
+using System.Runtime.CompilerServices;
+
+[assembly: AssemblyTitle("Microsoft.AspNet.SignalR.SqlServer")]
+[assembly: AssemblyDescription("Assembly containing components for scaling-out SignalR applications using SQL Server.")]
```
Comment (halter73) [inline src/Microsoft.AspNet.SignalR.SqlServer/Properties/AssemblyInfo.cs]: No need for the  `AssemblyTitle` or `AssemblyDescription` attributes anymore. The only thing you need is `InternalsVisableTo`.

File: .gitignore [hunk 7]
```diff
@@ -23,6 +23,5 @@ nuget.exe
 *.ncrunchsolution
 *.*sdf
 *.ipch
-/SignalR-Server.sln.ide
-
-*.sln.ide
\ No newline at end of file
+*.sln.ide
+debugSettings.json
```
Comment (halter73) [inline .gitignore]: I haven't seen this in any other .gitignore.

File: src/Microsoft.AspNet.SignalR.SqlServer/SqlServerServiceCollectionExtensions.cs (hunk 1 above)
Comment (Xiaohongt) [inline src/Microsoft.AspNet.SignalR.SqlServer/SqlServerServiceCollectionExtensions.cs]: Before I thought to change it in https://github.com/aspnet/SignalR-SqlServer/pull/2 after  https://github.com/aspnet/SignalR-Server/pull/65, now rename it.

File: src/Microsoft.AspNet.SignalR.SqlServer/TaskAsyncHelper.cs (hunk 2 above)
Comment (Xiaohongt) [inline src/Microsoft.AspNet.SignalR.SqlServer/TaskAsyncHelper.cs]: It is used, I thought to change to use shared class from Microsoft.AspNet.SignalR.Common after refactor.

File: src/Microsoft.AspNet.SignalR.SqlServer/SqlScaleoutConfiguration.cs (hunk 3 above)
Comment (Xiaohongt) [inline src/Microsoft.AspNet.SignalR.SqlServer/SqlScaleoutConfiguration.cs:12]: Currently ScaloutConfiguration is in Microsoft.AspNet.SignalR.Server, after refactor,  ScaloutConfiguration will be ScaloutOptions  in Microsoft.AspNet.SignalR.Messaging, so I thought to change SqlScaleoutConfiguration  after refactor.

File: src/Microsoft.AspNet.SignalR.SqlServer/TaskAsyncHelper.cs (hunk 2 above)
Comment (BrennanConroy) [inline src/Microsoft.AspNet.SignalR.SqlServer/TaskAsyncHelper.cs]: We want to get rid of it completely though. Starting with the scaleouts since they are easier and currently being worked on. There are 3 locations it is being used right now in this project. It would be great if we could remove them.

Comment (moozzyk) [inline src/Microsoft.AspNet.SignalR.SqlServer/TaskAsyncHelper.cs]: :+1:

File: .gitignore (hunk 7 above)
Comment (Xiaohongt) [inline .gitignore]: *.sln.ide is also in .gitignore in SignalR-Server

File: src/Microsoft.AspNet.SignalR.SqlServer/TaskAsyncHelper.cs (hunk 2 above)
Comment (Xiaohongt) [inline src/Microsoft.AspNet.SignalR.SqlServer/TaskAsyncHelper.cs]: Removed unused methods in TaskAsyncHelper.cs, but still have 20+ methods in TaskAsyncHelper.cs are used in SqlServer.

File: src/Microsoft.AspNet.SignalR.SqlServer/IDbCommandExtensions.cs [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+            if (sqlCommand != null)
+            {
+                return Task.Factory.FromAsync(
+                    (cb, state) => sqlCommand.BeginExecuteNonQuery(cb, state),
+                    iar => sqlCommand.EndExecuteNonQuery(iar),
+                    null);
+            }
+            else
+            {
+                return TaskAsyncHelper.FromResult(command.ExecuteNonQuery());
```
Comment (davidfowl) [inline src/Microsoft.AspNet.SignalR.SqlServer/IDbCommandExtensions.cs]: Change every use of TaskAsyncHelper.FromResult to Task.FromResult.

File: src/Microsoft.AspNet.SignalR.SqlServer/TaskAsyncHelper.cs (hunk 2 above)
Comment (halter73) [inline src/Microsoft.AspNet.SignalR.SqlServer/TaskAsyncHelper.cs]: The goal is to rewrite the few places that TaskAsyncHelper is used to use async/awit, `Task.FromResult`, etc... Then you can remove this class.

I don't care if you do this in a separate PR.

File: .gitignore (hunk 7 above)
Comment (BrennanConroy) [inline .gitignore]: debugSettings.json is a new one that is being recommended to add to the ignore file

**Comment by halter73** [comment]
:shipit:
